File: DSA/C++/Searching/DSA/binary_Search.py
# BinarySearch.py

def binary_search(arr, x):
    """
    Perform binary search on a sorted list 'arr' to find element 'x'.
    Returns the index if found, otherwise -1.
    """
    low, high = 0, len(arr) - 1

    while low <= high:
        mid = (low + high) // 2
        if arr[mid] == x:
            return mid
        elif arr[mid] < x:
            low = mid + 1
        else:
            high = mid - 1

    return -1


def main():
    print("=== Binary Search in Python ===")
    n = int(input("Enter number of elements: "))
    arr = list(map(int, input("Enter elements (sorted): ").split()))
    x = int(input("Enter element to search: "))

    result = binary_search(arr, x)

    if result != -1:
        print(f"Element found at index {result}")
    else:
        print("Element not found.")


if __name__ == "__main__":
    main()

File: DSA/C++/Searching/binarysearch.cpp

#include <bits/stdc++.h> 
using namespace std; 

int binarySearch(const vector<int>& arr, int target) {
    int left = 0;
    int right = arr.size() - 1;

    while (left <= right){
        int mid = left + (right - left) / 2;

        if (arr[mid] == target) {
            return mid;
        }
        if (arr[mid] < target) {
            left = mid + 1;
        }
        else {
            right = mid - 1;
        }
    }


    return -1;
}

int main() {
    vector<int> myVector = {2, 5, 8, 12, 16, 23, 38, 56, 72, 91};

    int targetToFind = 23;
    int result = binarySearch(myVector, targetToFind);

    if (result != -1) {
        cout << "Element " << targetToFind << " found at index " << result << endl;
    } else {
        cout << "Element " << targetToFind << " not found." << endl;
    }

    targetToFind = 44;
    result = binarySearch(myVector, targetToFind);

    if (result != -1) {
        cout << "Element " << targetToFind << " found at index " << result << endl;
    } else {
        cout << "Element " << targetToFind << " not found." << endl;
    }

    return 0;
}